src/base/ERC721Permit.sol

```diff
@@ -2,19 +2,79 @@
 pragma solidity ^0.8.24;
 
 import {ERC721} from "solmate/src/tokens/ERC721.sol";
+import {IERC1271} from "@openzeppelin/contracts/interfaces/IERC1271.sol";
+import {EIP712} from "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
 
-/// @notice An ERC721 contract that supports permit.
-/// TODO: Support permit.
-contract ERC721Permit is ERC721 {
-    constructor(string memory name_, string memory symbol_, string memory version_) ERC721(name_, symbol_) {}
+import {IERC721Permit} from "../interfaces/IERC721Permit.sol";
+import {UnorderedNonce} from "./UnorderedNonce.sol";
+
+/// @title ERC721 with permit
+/// @notice Nonfungible tokens that support an approve via signature, i.e. permit
+abstract contract ERC721Permit is ERC721, IERC721Permit, EIP712, UnorderedNonce {
+    /// @inheritdoc IERC721Permit
+    /// @dev Value is equal to keccak256("Permit(address spender,uint256 tokenId,uint256 nonce,uint256 deadline)");
+    bytes32 public constant override PERMIT_TYPEHASH =
+        0x49ecf333e5b8c95c40fdafc95c1ad136e8914a8fb55e9dc8bb01eaa83a2df9ad;
+
+    /// @notice Computes the nameHash and versionHash
+    constructor(string memory name_, string memory symbol_, string memory version_)
+        ERC721(name_, symbol_)
+        EIP712(name_, version_)
+    {}
+
+    /// @inheritdoc IERC721Permit
+    function DOMAIN_SEPARATOR() public view returns (bytes32) {
+        return _domainSeparatorV4();
+    }
+
+    /// @inheritdoc IERC721Permit
+    function permit(address spender, uint256 tokenId, uint256 deadline, uint256 nonce, uint8 v, bytes32 r, bytes32 s)
+        external
+        payable
+        override
+    {
+        if (block.timestamp > deadline) revert DeadlineExpired();
+
+        address owner = ownerOf(tokenId);
+        if (spender == owner) revert NoSelfPermit();
+
+        bytes32 digest = _hashTypedDataV4(keccak256(abi.encode(PERMIT_TYPEHASH, spender, tokenId, nonce, deadline)));
+
```

Comment:
I think we could really improve readability here a lot by moving this logic into a signature verification library..

really all this should look like in this code path is

```
bytes memory signature = ...
bytes32 permitHash = Permit.hash(spender, tokenId, nonce, deadline)
signature.verify(_hashTypedDataV4(permitHash), owner);  // also does all the support for IERC1271, and the ecrecover logic etc..
```

can reference also the permit2 signature library as well should be pretty similar